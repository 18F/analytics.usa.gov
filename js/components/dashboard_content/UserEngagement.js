import React from "react";
import PropTypes from "prop-types";

import EngagementRate from "./EngagementRate";
import AverageEngagementDuration from "./AverageEngagementDuration";
import FileDownloads from "./FileDownloads";
import ConsolidatedBarChart from "../chart/ConsolidatedBarChart";

/**
 * Contains charts and other data visualizations for the user engagement section
 * of the site. This component is mainly laying out the structure for the
 * section and passes props necessary for getting data and displaying
 * visualizations to child components.
 *
 * @param {object} props the properties for the component
 * @param {string} props.dataHrefBase the URL of the base location of the data
 * to be downloaded including the agency path. In production this is proxied and
 * redirected to the S3 bucket URL.
 * @returns {import('react').ReactElement} The rendered element
 */
function UserEngagement({ dataHrefBase }) {
  return (
    <>
      <div className="grid-row padding-top-3">
        <div className="desktop:grid-col-6">
          <div className="chart__title">User Engagement Events</div>
          <ConsolidatedBarChart
            dataUrl={`${dataHrefBase}/top-user-engagement-events-30-days.json`}
            chartDataKey={"event_label"}
            maxItems={10}
          />
        </div>
        <div className="desktop:grid-col-6 text--centered">
          <div>
            <AverageEngagementDuration dataHrefBase={dataHrefBase} />
          </div>
          <div className="padding-top-3">
            <EngagementRate dataHrefBase={dataHrefBase} />
          </div>
        </div>
      </div>
      <div className="grid-row">
        <div className="grid-col-12">
          <FileDownloads dataHrefBase={dataHrefBase} />
        </div>
      </div>
    </>
  );
}

UserEngagement.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
};

export default UserEngagement;
