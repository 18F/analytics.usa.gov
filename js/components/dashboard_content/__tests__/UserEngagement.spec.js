import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import UserEngagement from "../UserEngagement";

describe("UserEngagement", () => {
  let component;

  beforeEach(async () => {
    component = await render(
      <UserEngagement dataHrefBase="http://www.example.com/data/" />,
    );
    await waitFor(() => screen.getByText("Percent of Engaged Sessions"));
  });

  it("renders", () => {
    expect(component.asFragment()).toMatchSnapshot();
  });
});
