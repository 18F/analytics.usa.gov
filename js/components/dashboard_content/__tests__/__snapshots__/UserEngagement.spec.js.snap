// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserEngagement renders 1`] = `
<DocumentFragment>
  <div
    class="grid-row padding-top-3"
  >
    <div
      class="desktop:grid-col-6 text--centered"
      id="average_engagement_time"
    >
      <a
        class="chart__title"
        href="/definitions#report_historical_average_engagement_time"
      >
        Average Engagement Time Per Session
      </a>
      <div>
        <div
          class="data chart__rate padding-top-05"
        />
      </div>
    </div>
    <div
      class="desktop:grid-col-6 text--centered"
      id="engagement_rate"
    >
      <a
        class="chart__title"
        href="/definitions#report_historical_engagement_rate"
      >
        Percent of Engaged Sessions
      </a>
      <div>
        <div
          class="data chart__rate padding-top-05"
        />
      </div>
    </div>
  </div>
  <div
    class="grid-row"
  >
    <div
      class="grid-col-12"
    >
      <div
        class="padding-top-3 grid-row"
      >
        <div
          class="desktop:grid-col-6"
          id="file_extension_downloads"
        >
          <div
            class="text--centered"
          >
            <div
              class="chart__title"
            >
              <a
                href="/definitions#dimension_file_download"
              >
                Total File Downloads
              </a>
            </div>
            <div>
              <div
                class="data chart__rate padding-top-05"
              >
                ...
              </div>
            </div>
          </div>
          <div
            class="padding-top-3"
          >
            <div
              class="chart__title"
            >
              <a
                href="/definitions#dimension_file_extension"
              >
                Downloads by File Extension
              </a>
            </div>
            <figure
              id="chart_file_extension"
            >
              <div
                class="data chart__bar-chart"
              />
            </figure>
          </div>
        </div>
        <div
          class="desktop:grid-col-6 text--centered"
          id="file_download_counts"
        >
          <div
            class="chart__title"
          >
            <a
              href="/definitions#dimension_file_download"
            >
              Total Users with File Downloads
            </a>
          </div>
          <div>
            <div
              class="data chart__rate padding-top-05"
            >
              ...
            </div>
          </div>
          <div
            class="padding-top-9"
          >
            <div
              class="grid-row"
            >
              <div
                class="grid-col-6"
              >
                <figure
                  id="pie_chart_user_engagement_downloads"
                />
              </div>
              <div
                class="grid-col-6"
              >
                <div
                  class="chart__pie-chart__key"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
